Progress reporting improvements

dzenanz opened the pull request:
Related to #2875.

jhlegarreta reviewed (commented):
Have not tried, but wondering whether
> Now the progress fills up twice faster so it reaches 100% when only 50% of work is done.

is the expected behavior, or an expected limitation (e.g. if it cannot be set to 50% when half of the job has been done, and to be 100% only when the job has been completed), whether the wording is accurate or else whether it is a misunderstanding of mine.

blowekamp commented on Modules/Filtering/LabelMap/include/itkLabelMapToLabelImageFilter.hxx:
```diff
@@ -29,6 +29,7 @@ template <typename TInputImage, typename TOutputImage>
 LabelMapToLabelImageFilter<TInputImage, TOutputImage>::LabelMapToLabelImageFilter()
 {
   m_OutputImage = nullptr;
+  this->ThreaderUpdateProgressOff();
```
The base class `LabelMapFilter` is also disabling this. Is this really needed?

dzenanz replied:
Uh-oh. I didn't look at inheritance hierarchy.

I checked several children of `LabelMapFilter`, they all handle progress reporting themselves. I will remove this duplication.

blowekamp commented on Modules/Filtering/LabelMap/include/itkLabelShapeKeepNObjectsImageFilter.hxx:
```diff
@@ -29,6 +29,7 @@ LabelShapeKeepNObjectsImageFilter<TInputImage>::LabelShapeKeepNObjectsImageFilte
   m_NumberOfObjects = 1;
   m_ReverseOrdering = false;
   m_Attribute = LabelObjectType::NUMBER_OF_PIXELS;
+  this->ThreaderUpdateProgressOff();
```
LabelShaeKeepNObjectsImageFilter is also derived from InPlaceLabelMap/LabelMapFilter

dzenanz commented:
Before 840bae7 progress would go from 0% to 100% twice, overflowing from 100% to 0% about halfway. After 840bae7, the progress fills up to 100% during first half of the work, and stays there the second half of the work. Now, the progress fills up to 50% during first half of the work, and fills up to 100% during second half of the work.

In all instances, the progress was usually being filled smoothly.

blowekamp reviewed (approved):
Looks good. I just would like to know why it is needed to set is multiple times in a class hierarchy of constructors

dzenanz commented:
@jhlegarreta I updated the commit message.

jhlegarreta reviewed (approved):
Thanks for the patch set and the commit message update, now it makes more sense to me.

After Brad's comment, I'm wondering whether setting/updating the threader update progress report flag should be disabled in child classes/whether that makes sense.

dzenanz commented:
I have a more comprehensive solution. I am testing it locally.

thewtex commented:
@dzenanz please start commit messages with capital letters for the ChangeLog.

dzenanz commented:
There is something wrong with progress updates the second time filter is run:
```text
1570: Trying filter->UpdateLargestPossibleRegion()
1570: -------- Start GradientMagnitudeRecursiveGaussianImageFilter "" Progress Quiet
1570: Filter took 0.0594145 seconds.
1570: -------- End GradientMagnitudeRecursiveGaussianImageFilter ""
1570:
1570: itk::ExceptionObject (000000D8CD2FE3C8)
1570: Location: "void __cdecl itk::SimpleFilterWatcher::EndFilter(void)"
1570: File: C:\Dev\ITK-git\Modules\Core\Common\include\itkSimpleFilterWatcher.h
1570: Line: 283
1570: Description: ITK ERROR: GradientMagnitudeRecursiveGaussianImageFilter(000000D8CD2FED70): Filter does not have progress.
1570:
1570:
1570:   In C:\Dev\ITK-git\Modules\Filtering\ImageGradient\test\itkGradientMagnitudeRecursiveGaussianFilterTest.cxx, line 217
1/1 Test #1570: itkGradientMagnitudeRecursiveGaussianFilterTest ...***Failed    0.47 sec
```

dzenanz commented:
> something wrong with progress updates the second time filter is run

This is the only problem left. How much of a problem is this? Does someone know how to solve it? I am somewhat tired of this PR. It started as quite a small fix.

dzenanz commented:
I found out the problem. It will take me some time to clean up the code. I will push the update tomorrow.

blowekamp commented on Modules/Core/Common/include/itkProgressAccumulator.h:
```diff
@@ -138,6 +138,8 @@ class ITKCommon_EXPORT ProgressAccumulator : public Object
     unsigned long ProgressObserverTag;
     unsigned long StartObserverTag;
+
+    bool AccumulatedProgress{ 0.0 };
```
I am getting an error here with conversion from double to bool.

dzenanz replied:
I changed the variable name and how it is used in code, but forgot to change the type. And I was wondering why does debugger keep representing it as a float. Silly me! Thank you Brad.

blowekamp commented:
In testing this PR against SimpleITK I got a large number of FFT related test further inailures, and one Progress report test failing. It will need further investigation.

dzenanz commented:
Thanks for testing. Feel free to push to this this branch if you find some easy fixes.

dzenanz commented:
PythonLaplacianImageFilterTest deadlocks on my computer too. At least I can reproduce the failure.

dzenanz commented:
CI Linux Python:
```text
	209 - PythonGradientMagnitudeRecursiveGaussianImageFilterTest (Timeout)
	210 - itkGradientVectorFlowImageFilterPythonTest (Timeout)
	211 - PythonLaplacianImageFilterTest (Timeout)
	215 - PythonGeodesicActiveContourLeftVentricleTest (Timeout)
```

dzenanz commented:
I have removed the problematic commit from this PR. I will see if I can iron it out in a follow-up PR. I think this can be merged now.

dzenanz commented:
I removed all far-reaching code from this PR, and left all the non-problematic code here. It probably no longer fixes #2875, so I changed the description accordingly. I plan the real fix for a follow-up PR.

jhlegarreta reviewed (commented):
:+1: Thanks @dzenanz. Given that @blowekamp tested a previous version against SimpleITK, may be he would like to take that again, but I'm for merging it as the one-but-last build passed tests, and hope this will as well.

blowekamp commented:
Having to reset internal filter's progress when re-running seems like a red flag to me. I'd like to test with just re-running a filter in a standalone test to alleviate my concern. I'm not sure I'd have time tomorrow. There are a lot of changes here.

blowekamp reviewed (approved):
I tested a few cases that were of concerned to me, and they all behaved correctly 🥳 

Also SimpleITK tests seemed to be OK too (although there are other FFT issues).

Nice work!

dzenanz commented:
Are the FFT issues captured in #3048?

tbirdso commented:
> Also SimpleITK tests seemed to be OK too (although there are other FFT issues).

@blowekamp Could you please add @dzenanz  and myself to FFT discussion if there are other open issues?

blowekamp commented:
When I narrow in down in SimpleITK, I'll create an issue if needed. Here is a build with the failure FYI: https://open.cdash.org/viewTest.php?onlyfailed&buildid=7657537

tbirdso commented:
At a glance I'm guessing that the SimpleITK tests will need to add factory registrations for VNL FFT classes as the segfaulting tests all seem to require classes that depend on FFT filter instantiations. Whenever you get around to taking a closer look I'd recommend cross-referencing #3002 to see how this is being done in the ITK TestKernel as well as ITK GTests.
